server/grpc_bytestream.go

```diff
@@ -0,0 +1,361 @@
+		if err != nil {
+			recvResult <- status.Error(codes.Internal, err.Error())
+			return
+		}
+		resp.CommittedSize += int64(n)
+
+		for {
+			req, err := srv.Recv()
+			if err == io.EOF {
+				recvResult <- io.EOF
+				return
+			}
+			if err != nil {
+				recvResult <- status.Error(codes.Internal, err.Error())
+				return
+			}
+			if req.ResourceName != "" && resourceName != req.ResourceName {
+				msg := fmt.Sprintf("Resource name changed in a single Write %v -> %v",
+					resourceName, req.ResourceName)
+				recvResult <- status.Error(codes.InvalidArgument, msg)
+				return
+			}
+
+			n, err := pw.Write(req.Data)
+			if err != nil {
+				recvResult <- status.Error(codes.Internal, err.Error())
+				return
+			}
+			resp.CommittedSize += int64(n)
+
+			if resp.CommittedSize > size {
+				recvResult <- fmt.Errorf("Client sent more than %d data! %d",
+					resp.CommittedSize, size)
+				return
+			}
+
+			// Possibly redundant check, since we explicitly check for
+			// EOF at the start of each loop.
+			if req.FinishWrite {
+				recvResult <- io.EOF
+				return
+			}
+		}
+
+		if resp.CommittedSize != size {
+			err := fmt.Errorf("Unexpected amount of data read: %d expected: %d",
+				resp.CommittedSize, size)
+			recvResult <- status.Error(codes.Unknown, err.Error())
+			return
+		}
+
+		recvResult <- nil
+	}()
+
+	select {
+	case err, ok := <-recvResult:
+		if !ok {
+			msg := "Receive loop closed unexpectedly."
+			s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s", msg)
+			return status.Error(codes.Internal, msg)
+		}
+		if err == io.EOF {
+			pw.Close()
+			break
+		}
+		if err != nil {
+			s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s",
+				err.Error())
+			return err
+		}
+
+	case err, ok := <-putResult:
+		if !ok {
+			msg := "Cache Put closed unexpectedly."
+			s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s", msg)
+			return status.Error(codes.Internal, msg)
+		}
+		if err == nil {
+			// Unexpected early return. Should not happen.
+			s.accessLogger.Printf("GRPC BYTESTREAM WRITE CACHE INTERNAL ERROR")
+			return status.Error(codes.Internal, "Cache attempt failed.")
+		}
+
+		s.accessLogger.Printf("GRPC BYTESTREAM WRITE CACHE ERROR: %s",
+			err.Error())
+		return err
+	}
+
+	err, ok := <-putResult
+	if !ok {
+		msg := "Cache Put closed unexpectedly."
+		s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s", msg)
+		return status.Error(codes.Unknown, msg)
+	}
+	if err != nil {
+		s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s", err)
+		return status.Error(codes.Unknown, err.Error())
+	}
+
+	err = srv.SendAndClose(&resp)
+	if err != nil {
+		s.accessLogger.Printf("GRPC BYTESTREAM WRITE FAILED: %s", err)
+		return status.Error(codes.Unknown, err.Error())
+	}
+
+	s.accessLogger.Printf("GRPC BYTESTREAM WRITE COMPLETED: %s", resourceName)
+	return nil
+}
+
+func (s *grpcServer) QueryWriteStatus(context.Context, *bytestream.QueryWriteStatusRequest) (*bytestream.QueryWriteStatusResponse, error) {
```

Comment:
I meant that this can't operate on a partial/failed bs.Write call with bazel-remote specifically, because bazel-remote does not support partial/failed CAS writes (as far as I can tell it just deletes the data and the client has to start from the beginning).

Anyway updated to return 0, which I guess is OK for bazel and I suspect no other clients use this API.